RandomizerCore/StringParsing/Tokenizer.cs

```diff
@@ -0,0 +1,262 @@
+                    {
+                        TokenType = StructuralToken.Type.OpenParenthesis
+                    };
+                    AdvanceState();
+                }
+                else if (next == ')')
+                {
+                    AdvanceState();
+                    Consume();
+                    tokenInProgress = new StructuralToken
+                    {
+                        TokenType = StructuralToken.Type.CloseParenthesis
+                    };
+                    AdvanceState();
+                }
+                else if (stringDelimiter != null && next == stringDelimiter)
+                {
+                    tokenInProgress = ReadStringToken();
+                }
+                else if (reservedOperatorChars.Contains(next))
+                {
+                    tokenInProgress = ReadOperatorToken();
+                }
+                else
+                {
+                    tokenInProgress = ReadNameOrNumericToken();
+                }
+
+                // and just as we started, we end by eating excess whitespace
+                if (state != TokenizingState.TrailingTrivia)
+                {
+                    throw new TokenizingException($"Tokenizing in bad state - expected " +
+                        $"TrailingTrivia, was {state} at position {cursor}. " +
+                        $"Current partial token: {leadingTrivia}{content}{trailingTrivia}");
+                }
+                while (!IsEmpty() && char.IsWhiteSpace(Peek()))
+                {
+                    Consume();
+                }
+
+                tokenInProgress.StartCharacter = startingIndex;
+                tokenInProgress.EndCharacter = endingIndex;
+                tokenInProgress.LeadingTrivia = leadingTrivia.ToString();
+                tokenInProgress.TrailingTrivia = trailingTrivia.ToString();
+                tokens.Add(tokenInProgress);
+                StartToken();
+            }
+            return tokens;
+        }
+
+        // precondition: stringDelimiter must be nonnull
+        private Token ReadStringToken()
+        {
+            Expect(stringDelimiter.Value);
+            AdvanceState();
+            while (Peek() != stringDelimiter.Value)
+            {
+                Consume();
+                if (IsEmpty())
+                {
+                    throw new TokenizingException($"Encountered unterminated logic string starting at position {startingIndex}");
+                }
+            }
+            AdvanceState();
+            Expect(stringDelimiter.Value);
+            return new StringToken
+            {
+                Value = content.ToString(),
+            };
+        }
+
+        private Token ReadOperatorToken()
+        {
+            AdvanceState();
+            char next;
+            OperatorTokenizerTree tree = operatorTokenizerTree;
+            // this approach misses cases where operators are not leaf nodes. For example, if we define the operators >|> and *,
```

Comment:
Seeking feedback on this, not clear what the best approach would be to document this behavior in user-facing docs, or otherwise how to resolve this in a satisfactory way

Reply:
Obsolete by 37a05c00e9fe63e937f22d567e0ae22af78334b0